### ISSUE TITLE
Allow Multiple Classrooms Per Organization

### ISSUE THREAD

**Issue by tabinks** [education/classroom#595]
Now that GitHub has unlimited repositories for educational organizations (I'm making this assumption since my account was upgraded, although I never saw anything official from GitHub), it would be useful to have multiple classrooms per organization.  You are currently only allowed to have a single classroom per organization.

For example, if I had an organization named "University-CS101", I could have classrooms for "Fall-2015,Winter-2016,Spring-2016, etc.".  Previously, there could only be a single classroom for the organization and I would have to create organizations name "University-CS101-Fall-2015", etc.  This would allow educators to organize by the course.

**Comment by johndbritton** [comment]
> it would be useful to have multiple classrooms per organization.

I think this is something we want to support eventually. For now, you can create another organization and apply for the education program at https://education.github.com/discount to get free private repositories for your course.

We decided to limit it to one classroom per organization at first because it's easier to move to multiple classrooms per organization than it is to move the other way.

Would love to have your input on how this would work, what do you imagine the flow looking like? Can you use more than one organization per account? How do you group repositories from multiple courses in one organization?

**Comment by drbeco** [comment]
I support this idea. For example, one organization should be enough to put together diverse classrooms from one teacher (I teach 4 courses, -- each semester --).

Also, if we can have multiple owners in one organization, we can set other professors to be part of it. Then we would actually have the hierarchy:

Organization: University Foobar
Team1: Professors
Team2...N: different groups of students from the same class, or different classes, or a subset for each task and so on.
Classrooms: each professor can create several within the organization
Assignments: each classrooms can have multiples assignments, each one with a different team, that can or cannot intersect with others.

I'm new to these business (github education and classroom) but I like the way it is going, and I hope we will have more freedom to organize our stuff.

Thank you all!

**Comment by johndbritton** [comment]
I've been thinking about this. What if we changed the mapping in Classroom to just be 1:1 organization with organization and within an organization in Classroom you could group assignments. Each group of assignments would be one course but all the repositories would live in one organization on GitHub.

One thing that complicates multiple classrooms in one organization is teams. You can't have two teams with the same name in two classrooms if they're both in the same GitHub organization.

A way around that would be to have a top level organization that has teams and let teachers group assignments. In Classroom you would have one organization for all of your courses, but you could look at each course independently even though some things are shared amongst the courses (like teams).

**Comment by johndbritton** [comment]
@tarebyte what do you think about this?

**Comment by drbeco** [comment]
I think I see your problem with the relational database.

Suppose you set 1:1 classroom:organization. Ok, first of all, classroom will then be a "feature" of an organization. That is ok. Just change the way it is promoted and explained.

Now, if we can "group" assignments, we might have a problem with nomenclature. Maybe, if possible, we should name this groups "classrooms". Although the "Application" itself is named classroom and refers to the (now) feature, that is just technical talking. No one needs to know these details except the developers.

So if it is ok for you, we would have groups of assignments named "classrooms". This would take the burden of splitting a higher level of code.

Regarding the team, I don't see why someone needs two teams with the same name. Just name it something else. No problem here.

The only thing that needs to be compartmentalized is the assignment. The rest is just levels of hierarchy to help us organize it. And if someone is in two teams for two assignments, well, that is ok he/she can read both.

I may be missing some more important problem you are facing.

**Comment by johndbritton** [comment]
> Now, if we can "group" assignments, we might have a problem with nomenclature. Maybe, if possible, we should name this groups "classrooms". Although the "Application" itself is named classroom and refers to the (now) feature, that is just technical talking. No one needs to know these details except the developers.

Yeah, I agree with you here. This should be something that we make really simple for the end user.

> The only thing that needs to be compartmentalized is the assignment.

Organization membership is another thing that will be polluted if we have multiple classrooms in one org. I don't think it's a blocker, but something that we need to think about.

This is something I think we will do, but it's likely to come after more important features like due dates, assignment submission, invitations, and team management.

**Comment by tabinks** [comment]
>Would love to have your input on how this would work, what do you imagine the flow looking like? Can you use more than one organization per account? How do you group repositories from multiple courses in one organization?

Here is the flow I would imagine for an example course, CS101, that I teach multiple times per year:
* I have an existing organization named "CS101" 
* I create a new "Classroom" in the web application named "Spring-2016"
* Later to create an assignment, I go to the Classroom web app, choose my classroom and create a new assignment named "Assignment1-HelloWorld"
* Students receive an invitation for the assignment.  The assignment is automatically named "CS-101_Spring-2016_Assignent1-HelloWorld_githubStudnetUserName".  The key here is the we are appending the classroom name. 
* During next quarter, I create a new classroom named "Summer-2016" and so on.

This is basically what I'm doing manually, but before the unlimited repo option, I added the step to delete previously  quarter assignments (when I was near the quota).

> What if we changed the mapping in Classroom to just be 1:1 organization with organization and within an organization in Classroom you could group assignments. Each group of assignments would be one course but all the repositories would live in one organization on GitHub.

This would work out.  Basically, what I would rather not have to do is create three organizations: CS101-Autum, CS101-Winter, CS101-Spring.  Having all repos live under the same organization is fine, the grouping of classrooms would just be something to better help manage the GitHub Classroom.


> One thing that complicates multiple classrooms in one organization is teams. You can't have two teams with the same name in two classrooms if they're both in the same GitHub organization.

Personally, my courses do not take advantage of teams as the assignments are mostly individual.  I usually create a team for each course so that I can "@" the entire course in Issues.  We use Issues as a class forum.  This would not be an issue for us.

> A way around that would be to have a top level organization that has teams and let teachers group assignments. In Classroom you would have one organization for all of your courses, but you could look at each course independently even though some things are shared amongst the courses (like teams).

This sounds good.

> This is something I think we will do, but it's likely to come after more important features like due dates, assignment submission, invitations, and team management.

Just my 2 cents, but we've never really had issues invitations or team management.  They work fine as is.  The due dates and assignment submission are nice thoughts, but this hasn't been an issues.   We tell students that `master` is their submission repo and we will grade the last repo in this branch before the deadline has passed.  I've thought of making better use of git tags for this, but, in practice, this solution works fine.

**Comment by johndbritton** [comment]
> Basically, what I would rather not have to do is create multiple organizations (eg. CS101-Autum, CS101-Winter, CS101-Spring). Having all repos live under the same organization is fine, the grouping of classrooms would just be something to better help manage the GitHub Classroom.

This is really good feedback. Thanks for sharing.

> The due dates and assignment submission are nice thoughts, but this hasn't been an issue either. We tell students that master is their submission repo and we will grade the last repo in this branch before the deadline has passed. I've thought of making better use of git tags for this, but, in practice, this solution works fine.

Classroom works as is so long as you can enforce your rules using convention. For larger courses or with really newbie students it's hard to get them to understand how submission works. Due dates and submissions have been our most requested features.

Adding team management will let teachers self-service and fix cases where students accidentally join the wrong team. Right now there's no way to fix that short of emailing GitHub support.

**Comment by tabinks** [comment]
Another feature that may be useful would be the ability to "archive" a course.  For example, once CS101-Autumn is completed, provide some way to lock the repos and potentially hide them (or provide some other visual cue that the course is over).

**Comment by johndbritton** [comment]
> the ability to "archive" a course.

@tabinks Would you be willing to open another issue for that feature request and provide some ideas / reasoning for it so that it doesn't get lost in the shuffle here?

**Comment by tabinks** [comment]
@johndbritton I have created a new issue #602.

**Comment by zachfeldman** [comment]
This is a SUPER important feature for many schools to be able to use this platform at all. When we saw what was going on with this feature from GitHub here @nycda we were so excited! But once we realized we wouldn't be able to break up our various classes under our one organization, we were sad that we wouldn't be able to use . We'll see if we can divert the resources to hack something up but as a built-in feature, having multiple classrooms in the same org would be a lifesaver.

**Comment by vbrh-immalle** [comment]
I was also surprised to see this was not possible. For now, I will try to seperate my two classes by using (secret) teams and see how it goes...

**Comment by dhowe** [comment]
+1

**Comment by csteacherd22** [comment]
Agreed. More classrooms per organization is crucial - anything else is basically a kludge.
What is the status on this feature request?

**Comment by webdif** [comment]
Is there any progress about this issue? It's a critical issue for my organization. Thanks!

**Comment by tarebyte** [comment]
@webdif unfortunately right now I just don't have the bandwidth to work on it. If anyone has the time to make it happen I'd be happy to review the work.

**Comment by jaegernh** [comment]
I just stumbled upon this and agree with it's importance. I might take a look if I can find the time. @tarebyte

**Comment by KokoDoko** [comment]
Agreed. If the organisation is a school then it makes sense that there can be more than 1 classroom! It's awkward to create a new organisation for every class or every semester!

**Comment by paragmantri** [comment]
Is here any update on the issue. As I see it now, there is still only one classroom per organisation.

**Comment by nwoodthorpe** [comment]
It seems a lot of people are interested in this feature. We still have some implementation details that need to be ironed out, but I'll try to prioritize solving this.

**Comment by paragmantri** [comment]
Thanks @nwooodtorpe. Once you do that, I hope assignment submissions will appear under each classroom. As of now, all the student submissions appear directly under the organisation and the classroom is not in the tree. It can get very confusing to differentiate between our own repo as well as the ones submitted by students.

**Comment by jfiksel** [comment]
I would also like to recommend this feature. It seems like one could easily get bogged down with added a new organization for each semester. In addition, it would be nice to have generic starting code for different assignments within an organization, and then when creating new classrooms for each semester, the starter code could then be re-used very easily.

**Comment by gregnordstrom** [comment]
I'm another teacher in support of multiple classrooms per organization. It seems to map very well to the way I and others logically group our teaching material. Kudos to @tabinks who captured it quite well. Also, @webdif mentioned the desire to create assignments with starter code from another organization's private repo (e.g. a "master collection" of material for a given course that I maintain from year to year). Another great idea! Thanks for listening and for the great support!

**Comment by brks** [comment]
@nwoodthorpe Hi, we're about to start our projects for 2017/18 and would like to use GitHub Classroom although having Multiple Classrooms Per Organisation is going to be required for us to use it. Any idea when you'll have time to work on this?

**Comment by nwoodthorpe** [comment]
@tarebyte would this be do-able by namespacing assignment repos with the classroom name and namespacing team names with the classroom name?

**Comment by tarebyte** [comment]
@nwoodthorpe let me think it over, but that's sounds about right.

**Comment by BrianEmilius** [comment]
I'm a teacher and I wholeheartedly support this feature.
It would reduce a lot of noise to be able to create multiple classrooms within an organization, it's intuitive, and it just makes plain and simple sense.

**Comment by brks** [comment]
@tarebyte @nwoodthorpe any developments with this feature?

**Comment by mozzadrella** [comment]
@brks 👋  saw you signed up to user-test new features :) Excellent, thank you.
This feature is not part of our current theme (https://github.com/education/classroom/projects/7) 
But it may be included in our next theme (~July-August)

**Comment by c4cheats** [comment]
Just wanted to voice my support for this feature as well.

**Comment by Garylum** [comment]
I'd also like to voice support for multiple classrooms per org as well.  Another sticking point for us is the owner requirement to be able to admin a classroom.  We'd like to give each instructor admin rights to their class but not have to give them owner rights to our whole organization if possible

**Comment by mozzadrella** [comment]
@Garylum @c4cheats thank you both for voicing your support, when we begin our next sprint this is part of the "Manage and Assess" theme we want to take on: https://github.com/education/classroom/milestone/5

**Comment by msraunak** [comment]
I would like to also voice my support for multiple classrooms per organization. I will be teaching three different courses this Fall and I plan to use Github classroom in all three courses. It will be a whole lot more convenient to have all three classes under one organization.

**Comment by readoc** [comment]
I'm all for this as well. I would rather keep all classes under one organization if possible. So, +1

**Comment by Badlapje** [comment]
I definitely support this.  

From a UX point of view, i find the current structure to be very unclear, as you're using the term organization to mean course.  From an administrative point of view, you've created quite a bit of overhead. Right now i need to create an organization not only for each course, but also for each class in that course.  I teach 10 courses / year.  If i want to put all of them on github classes, i'll have to send in 10 tickets / year + create 10 organizations + create 10 classrooms.  And the year after that i can start again.  My collegues will have to do the same, so you will end up with 60 tickets / year, from one department in my school.   

The flow as i see it would be this:
1. create one organization.  An organization i imagine as being one of two things:
  a. a school / university / college / educational organization.  Eg. University of Stanford
  b. a department from a school / university / college / educational organization.  Eg. Stanford ACS
2. add a new structure: course.  Eg. CS101
3. within a course one can create as many classrooms as needed.  Eg. CS101-Fall-2017

Compare this with the current flow you'd reduce to overhead for my school from 60 tickets / year to just a single 1.  Once your organization is approved, it's all good and we can create the courses & classrooms we need.  Easy and done.

**Comment by Badlapje** [comment]
@tarebyte you mention we can help with the implementation of this feature?  Can you direct me to the correct repo to do that @?  I would be very happy to help with that.

**Comment by JamesSaxon** [comment]
I would also appreciate the ability to have multiple classrooms per organization.  It is the same class I'm teaching last year, so many of the resources are the same but many have also evolved.  The assignments are not all the same.

Having multiple classrooms would allow the course to evolve over time in a natural way.  

Moreover, I think this is an important part of the GitHub education philosophy -- namely, supporting students to build up a code history.  If I delete old assignments, it will delete the old students' work:

> This will permanently delete the hw-XXX assignment from GitHub Classroom. In addition this will also delete 57 participant repositories under the YY-ZZ organization.

On the other hand, _not_ deleting would confuse current students over old assignments in the classroom.

I strongly support the "archive" feature for finished classes.  It's even important, in case someone ends up asking/complaining about some grade _way_ late...

**Comment by jfiksel** [comment]
I also want to add my support of @Badlapje 's vision of GitHub Classroom organization.

**Comment by wschilling** [comment]
I'm curios as to the status of this change.  For accreditation purposes, it would be much easier if I could have a second classroom now for a second set of courses, as I am tag teaching with different people now.  But, I'd rather not have to have another organization, as really, only the class is different.  Any updates on this one?

**Comment by dcoreyrasm** [comment]
Just curious if there have been any updates on this? I'm currently writing a course and without the ability to add multiple classrooms with unlimited private repositories, it looks like I'll have to advise faculty to create their own accounts & organizations, then create a classroom...which I really wish would allow the unlimited private repositories without submitting a new request. (or am I wrong about this? Please let me know. I'm trying to streamline the process as much as I can without faculty having to create new organizations.)

My vision is similar to previous posts...
1 CIS department org, for example
with multiple classrooms which I would name using our course numbers and terms

**Comment by itggot-daniel-berg** [comment]
Hope this gets enabled soon. Would really, _really_ simplify using Github Classroom for my different courses.

**Comment by alhvi** [comment]
I'm also interested on this feature

**Comment by diegoep** [comment]
+1

**Comment by shivagoud** [comment]
I'd also like to voice support for multiple classrooms per org as well. I am an educator. I have been using classroom in a different way than intended.
- 1 organization and associated classroom has all the assignments.
- All assignments are group assignments
- Everytime a new batch of students join the course, I create a new team for them in the existing assignment.
- Each submission has to be in a unique directory so that all the submissions are merged.
- The last commit on the `master` branch is accepted as the submission.

Mine is an open organization where we allow each other to see and review the code. So there's no issues so far. But when I want to hide the assignment submissions from each other, then it is not possible with the current structure.

**Comment by mverdicchio** [comment]
Jumping on the bandwagon! Please create this feature!

**Comment by lwasser** [comment]
me too! i manage many classes and having many organizations will become very tedious. quickly

**Comment by ejgutierrez74** [comment]
+1 I also want to add my support of @Badlapje 's vision of GitHub Classroom organization.
Is there any updates/news improvements in this request ?

**Comment by Magistern** [comment]
I just want to add my voice to this matter. We have ben using GitLab on a self hosted server before and there you organize your "classes" using projects. 

I second the idea of having 1 organization (the school)
...having multiple teams/departments/subjects/professors
...having multiple classrooms
...having multiple annual versions.
...having multiple assignments.

That way, We would have the organization 
* Jönköping University
  * Eve
  * Phil
  * John
    * 2017
      * Client-side programming 17
        * Lab 1
        * Lab 2
        * Lab 3
      * Server-side programming 17
        * Lab 1
        * Lab 2
        * Group Project Assignment
    * 2018
      * Client-side programming 18
        * Lab 1
        * Lab 2
        * Lab 3

I guess everything would be easily solvable with the addition of Custom Hierarchical Taxonomies
You could name them whatever you like (school/class/professor etc) and create your own parent-child relationship. Then all you have to do is categorize each assignment using those taxonomies. 
Visually it could look like folders for instance.

**Comment by ejgutierrez74** [comment]
I think it would be clearer:

Jönköping University ( organization)
        Client-side programming 17 ( subject with year or course 18/19)
          Optionally this course/subject could have one or many classrooms ( ex: one in the mornings, one in the afternoon). Perhaps a university course with 200 students, could be divided in 4 clasrooms with 50 students each and with teachers assgined to each classroom. So classroom1 would have Ann and Peter as teachers, classroom2 with Joseph and classroom3 Teddy and Vanessa, classroom4 Eugene, Brad and Angeline. Every classroom has different students.
               Classroom 1 : ( has its own teachers, students and deadlines to each assignment)
                     Lab 1 ( assginement1)
                     Lab 2 ( assginement2)
                     Lab 3 ( assginement3)
               Classroom 2 : ( has its own teachers, students and deadlines to each assignment)
                     Lab 1 ( assginement1)
                     Lab 2 ( assginement2)
.....
              Classroom 4 : ( has its own teachers, students and deadlines to each assignment)
                     Lab 1 ( assginement1)
                     Lab 2 ( assginement2)
                     Lab 3 ( assginement3)
                     Lab 4 ( assginement4)
                
         Server-side programming 17
                ....
            Client-side programming 18
                Lab 1
                Lab 2
                Lab 3 
                Lab4

My approach i think is clearer because in one organization the course is done every year or semester, and teachers can move to another organization, or have different courses/subjects from one year to another. Persons change course remains..

**Comment by Magistern** [comment]
Sure, as I suggested, if we could just create and organize our categories and their hierarchical relationships our selves, we could create a structure that fits each organization, big or small.

So, taxonomies would be easy to implement, no fuzz about how it should be organized. Everyone does it their way.

GitHub already have a non hierarchical taxonomy structure; Topics, so implementing it
Would be a relatively  small adjustment
Johan

________________________________
Från: Eduardo Gutierrez <notifications@github.com>
Skickat: onsdag, september 12, 2018 4:56 em
Till: education/classroom
Kopia: Johan Kohlin; Comment
Ämne: Re: [education/classroom] Allow Multiple Classrooms Per Organization (#595)


I think it would be clearer:

Jönköping University ( organization)
Client-side programming 17 ( subject with year or course 18/19)
Optionally this course/subject could have one or many classrooms ( ex: one in the mornings, one in the afternoon). Perhaps a university course with 200 students, could be divided in 4 clasrooms with 50 students each and with teachers assgined to each classroom. So classroom1 would have Ann and Peter as teachers, classroom2 with Joseph and classroom3 Teddy and Vanessa, classroom4 Eugene, Brad and Angeline. Every classroom has different students.
Classroom 1 : ( has its own teachers, students and deadlines to each assignment)
Lab 1 ( assginement1)
Lab 2 ( assginement2)
Lab 3 ( assginement3)
Classroom 2 : ( has its own teachers, students and deadlines to each assignment)
Lab 1 ( assginement1)
Lab 2 ( assginement2)
.....
Classroom 4 : ( has its own teachers, students and deadlines to each assignment)
Lab 1 ( assginement1)
Lab 2 ( assginement2)
Lab 3 ( assginement3)
Lab 4 ( assginement4)

     Server-side programming 17            ....        Client-side programming 18            Lab 1            Lab 2            Lab 3             Lab4

My approach i think is clearer because in one organization the course is done every year or semester, and teachers can move to another organization, or have different courses/subjects from one year to another. Persons change course remains..

**Comment by jfiksel** [comment]
I agree with Johan here, this seems like a reasonable way to organize this. The way we've been getting around this is by having a master organization ([described here](https://github.com/jfiksel/github-classroom-for-teachers)) for a given class that has all of the base assignments, and then re-using or adding to these assignments for each new class (organization). So in addition to what Johan has suggested, it would be nice if we could easily re-use materials (starter repos & lecture repos, if that exists) from class-to-class without also importing all the student homework repos

**Comment by Magistern** [comment]
@srinjoym I see you closed and opened this request, does that mean you are considering it?

**Comment by ADWolfe-LoyNO** [comment]
PLEASE IMPLEMENT THIS REQUESTS FOR MULTIPLE CLASSROOMS PER ORGANIZATION

**Comment by lwasser** [comment]
woo hoo!! excited to hear this @srinjoym  !! thank you for the update on this issue.

**Comment by Magistern** [comment]
That's wonderful news! Looking forward to try it out.

**Comment by BrianEmilius** [comment]
> @Magistern We have been running a public beta of adding support for multiple classrooms in a GitHub Organization for the last month to ensure the feature works as intended. We're planning to release the feature to everyone later this week, along with a blog post on how to use it. Thanks for your patience raised_hands

This is fantastic news. Thanks for listening to the community!

**Comment by ADWolfe-LoyNO** [comment]
@srinjoym any further news on this?  We are starting to create assignments for upcoming classes and creating new orgs for them is just not a good situation

**Comment by mozzadrella** [comment]
@ADWolfe-LoyNO indeed! We now offer multiple classrooms per organization: 

https://github.blog/2019-01-23-github-classroom-now-supports-multiple-classrooms-per-organization/

Thanks to all who were patient with us as we worked on this feature. If you have feedback, feel free to reach out to education@github.com and put my handle somewhere in the message. 

Wishing you all an awesome semester. ✨

**Comment by ADWolfe-LoyNO** [comment]
Hooray!

Andrew D Wolfe, Jr, MSCS
Assistant Visiting Professor
Department of Mathematics and Computer Science
Loyola University of New Orleans
adwolfe@loyno.edu <mailto:adwolfe@loyno.edu>
tel:+15048652951

**Comment by jfiksel** [comment]
It's a bit unclear from the blog post how this works--does this mean that students in different classrooms will all be part of the same organization? And then the assignment repos will continue to all be within the same organization?

**Comment by uvic-frodo** [comment]
@mozzadrella With this new feature, the repository name does not include the classroom ID.  In particular, the repository name appears to be formed by the assignment ID + "-" + user ID.  This choice of naming convention then requires a scheme for avoiding name collisions when more than one repository is associated with the same assignment-ID student-ID pair (which can arise in some use cases).  As best I can tell, the name-collision avoidance scheme used is to append a "-" and an integer to the end of the repository name.  This begs the question:   Given a classroom ID, assignment ID, and user ID, how can an EXTERNAL tool determine the repository name?  This is a problem for me because I use external tools that break, due to being unable to deduce this information.  Is it not possible to name the repository something like: classroom ID + "-" + assignment ID + "-" + user ID?  This would help to avoid many problems for external tools.  Failing that, how can an external tool map a classroom ID, assignment ID, and user ID into a repository name?

**Comment by Kevin-CodeCrew** [comment]
> It's a bit unclear from the blog post how this works--does this mean that students in different classrooms will all be part of the same organization? And then the assignment repos will continue to all be within the same organization?

That's been my experience. Great, I can now have multiple classrooms, but individual assignment repos from all Students show up in our primary organization which will not do. We quickly would be buried in a list of 100s and eventually, 1000s of Student assignment entries. Our main organization was/is intended to be a place where Instructor materials are housed. Not individual Student assignments.   

I thought I would take another crack at setting something up as thought that I maybe missed something but no. I can find no way around this behavior. We really need private repos for assignments, but for now, will just have to stick to the 'create a new org per semester/semester/session' approach as that's worked well for the last 3 cohorts over 18 months.

Github kindly granted unlimited repos for our non-profit, but only for the main organization. They denied followup requests for cohort/semester/session-specific organizations saying that now that our org can have unlimited classrooms, multiple orgs should not be required.

If anyone has any tips/workarounds we might want to try so we can take advantage of our private repos for assignments I'd love to hear them. Thanks!
